File: 246.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

class Solution {
public:
    int nthUglyNumber(int n) {

        vector<long long> dp(n);
        dp[0] = 1;
        int twoPtr = 0;
        int threePtr = 0;
        int fivePtr = 0;
        long long twoMul = dp[twoPtr] * 2;
        long long threeMul = dp[threePtr] * 3;
        long long fiveMul = dp[fivePtr] * 5;

        for(int i=1;i<n;i++){
            dp[i] = min({twoMul, threeMul, fiveMul});
            if(dp[i] == twoMul)
                twoMul = dp[++twoPtr] * 2;
            if(dp[i] == threeMul)
                threeMul = dp[++threePtr] * 3;
            if(dp[i] ==  fiveMul)
                fiveMul = dp[++fivePtr] * 5;
        }
        return dp[n-1];
    }
};

int main() {
    Solution s;
    int n;
    cin >> n;
    cout << s.nthUglyNumber(n) << endl;
    return 0;
}

File: python/246.py
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        dp = [1] * n
        two_ptr, three_ptr, five_ptr = 0, 0, 0
        two_mul, three_mul, five_mul = 2, 3, 5
        for i in range(1, n):
            dp[i] = min(two_mul, three_mul, five_mul)
            if dp[i] == two_mul:
                two_ptr += 1
                two_mul = dp[two_ptr] * 2
            if dp[i] == three_mul:
                three_ptr += 1
                three_mul = dp[three_ptr] * 3
            if dp[i] == five_mul:
                five_ptr += 1
                five_mul = dp[five_ptr] * 5
        return dp[-1]
    
if __name__ == '__main__':
    s = Solution()
    print(s.nthUglyNumber(int(input())))